### ISSUE TITLE
Pixelate Effect Bug

### ISSUE THREAD

**Issue by peanutbutterandcrackers** [OpenShot/libopenshot#121]
I am on the latest daily builds from the develop branch and it seems that the 'Pixelate' video effect isn't really working. Perhaps it needs a little more testing (maybe it's just happening to my instance) but the preview isn't updated when the effect is applied and neither does the final render show anything pixelated. So, perhaps it is bugged?

**Comment by N3WWN** [comment]
What values are you using with the `Pixelate` effect?

The default (on my current dev fork, which is probably pre-develop branch and pre-rc) is 0.70 Pixelization and is a light pixelization.  0.30 is just barely pixelated _if_ you're looking for it.  0.99 is totally pixelated (about 12x7 squares filling the 720p preview).  

Values over 0.99 don't appear to work (1.0 is the same as no pixelization) which kinda makes sense cuz how much more pixelated do you want than 12x7 squares? 😜

**Comment by peanutbutterandcrackers** [comment]
Ah... I see. So this issue is just a false alarm.

Could you please explain why 0.99 works and 1 doesn't again? Does not quite ring a bell in my brain (I'm not that smart, TBH :D)

**Comment by N3WWN** [comment]
Looking at `src/effects/Pixelate.cpp`, the `Pixelization` value is essentially a percentage used to determine how much smaller the internal image buffer needs to be in order to pixelate the frame.  After the lower resolution internal buffer is created, the frame is copied into the buffer and then expanded again without smoothing (it's the "without smoothing" that is the secret sauce in this recipe) to create the full size, but pixelated frame.

Some examples:

> Original frame:  1920x1080 (100%)
Pixelization value: 0.7 (70%)
Internal buffer: 576x324 (100-70=30% of 1920x1280)

> Original frame:  1280x720 (100%)
Pixelization value: 0.7 (70%)
Internal buffer: 384x216 (100-70=30% of 1280x720)

> Original frame:  1280x720 (100%)
Pixelization value: 0.3 (30%)
Internal buffer: 896x504 (100-30=70% of 1280x720)

If we set the `Pixelization` value to 100%, it no longer works:

> Original frame:  1280x720 (100%)
Pixelization value: 1.0 (100%)
Internal buffer: 0x0 (100-100=0% of 1280x720)

The maximum value that libopenshot will accept is [0.9999](https://github.com/OpenShot/libopenshot/blob/d8aa28ba6e3a4a2987286fe6c619ba02b9c92afa/src/effects/Pixelate.cpp#L187), giving us:

> Original frame:  1920x1080 (100%)
Pixelization value: 0.9999 (99.99%)
Internal buffer: 0.192x0.108 (100-99.99=0.01% of 1920x1280)

or

> Original frame:  1280x720 (100%)
Pixelization value: 0.9999 (99.99%)
Internal buffer: 0.128x0.072 (100-99.99=0.01% of 1920x1280)

So, I tested (in my earlier post above) with 0.99 and saw a 12x7 pixel image.  Let's see if that works out:

> Original frame:  1280x720 (100%)
Pixelization value: 0.99 (99%)
Internal buffer: 12.8x7.2 (100-99=1% of 1920x1280)

Yup!

Does that clear things up? 😃

**Comment by peanutbutterandcrackers** [comment]
It does. Then, perhaps, we should make the possible values go from 0.0 to 0.99 only and not allow values like 1 and upwards be put in - only confusing the user. Don't you think?

**Comment by N3WWN** [comment]
There are lots of "issues" (well, not really issues, but idiosyncrasies) with how the properties are handled.  

Folks have reported problems with selected properties highlighting with the wrong color based on whether the property is keyframed, whether the property on _this_ frame is a keyframe, what the units are for each property, etc.

To implement what you're suggesting should be applied to all properties (clip, effect, transition, etc) across the board, which means we need a consistent way to handle such things.   

Do we simply limit the input values and ceil()/floor() them to the max or min values if the user enters values outside of the usable range?

Do we notify the user that the value has been adjusted?

Do we refuse to accept values that are out of range and revert to the previous value?

Since we should tackle units at the same time, do we allow users to specify the units they want to use for properties with multiple valid unit systems?  Think of the `Position` property... should it be frames or time?  Since it's currently time, but you may want to specify the position in frames, should we let folks enter '200f' or 'f200' or '1min2sec' or '00h:01m:02s' or '00:01:02' or '00:00:00s3f' or whatever to specify that you want to specify the item in one (or more) units?  

Do we have a per-property unit setting to change between unit systems?

If we do that, do we change the preferred unit based on the last entered value+unit or require users to right-click and select a unit or set the unit via the Preferences?

There's a lot of questions to ask (and answers to agree upon) to implement this correctly.

All of this would also be performed in openshot-qt, not here in libopenshot.

I think we should add a future feature request for all of the above, but only after we get the keyframe vs frame rate issue resolved so we aren't trying to implement multiple unit systems in a dynamic unit environment.

**Comment by peanutbutterandcrackers** [comment]
Hmm... True.

I had seen that if I give invalid input to the property, say an alphabet instead of a numeral, it doesn't accept it (silently), so I thought if similar thing could happen to this one, too.

But you're right. This needs to be discussed at length and agreed upon after the frame-rate vs keyframe issue is resolved. Which brings me to [this](https://github.com/OpenShot/libopenshot/tree/experimental/auto-convert-keyframes). Have you taken a look at that, good sir? Perhaps you could help out!

**Comment by peanutbutterandcrackers** [comment]
I am closing this issue for now as it is not really a bug (my bad).

**Comment by jeffski** [comment]
Just my 2 cents - I have started running in to issues around some things being time based and others frame based. I think the library should work in frames for everything and make it the responsibility of the application using the library to manage the conversion to/from time if it's needed.

I think frame accuracy needs to be a top priority, there are a couple of issues (one I posted) over the last few weeks where frame accuracy isn't correct which really makes it very difficult to create a quality, reliable and predictable render. See https://github.com/OpenShot/libopenshot/issues/115 and https://github.com/OpenShot/libopenshot/issues/122.

**Comment by peanutbutterandcrackers** [comment]
@jeffski - Perhaps you'd like to open up another issue for the **library to work in frames for everything**? Maybe that'd open up the much-needed discussion and other users could also put their views forward? (I still need to file that another bug report though (which is almost certainly related with the issues that you've mentioned - somehow.)
